# 📗 Do it! 자료구조 · 알고리즘 — **1일차 스터디 노트북**

CReAL 연구실 하계 자료구조 스터디 · 1일차 복습용

---

### 🚦 난이도 티어 (문제마다 표시)
| 태그 | 뜻 | 대상 |
|---|---|---|
| 🟢 **기본** | 오늘 개념을 **쓸 수 있는지** 확인 | **전원 필수** |
| 🟡 **표준** | **왜 그런지** 이해 + 변형·디버깅 | 팀 목표선 |
| 🔴 **심화** | 효율 분석 · 엣지케이스 · 일반화 | 도전 (여유 되면) |

> 📌 **스터디 운영 팁**: 🟢🟡는 다 같이 풀고, 🔴는 "풀 사람만". 금요일 코딩테스트는 🟢🟡 범위에서.
> 막히면 바로 물어보기 — 혼자 30분 붙잡지 말 것.

### 🧭 푸는 법
- **[예측]** 표시: 실행하기 **전에** 결과를 주석/머릿속으로 먼저 적고 → 실행해서 맞춰보기.
- **[구현]** 표시: `# TODO` 자리를 직접 채우기. 기대 출력이 적혀 있으니 스스로 검증 가능.
- **[설명]** 표시: 말이나 글로 "왜?"를 설명해보기.
- 다 풀면 맨 아래 **✅ 정답 & 해설** 에서 채점. (먼저 풀고 보기!)


### 🗺️ 1일차 개념 지도
> 파이썬 문법은 안다고 가정. 오늘은 **알고리즘적 사고**에 초점.

- **변수·타입**: 값을 담는 그릇 · `input()`은 항상 `str` → `int()` 형변환
- **연산자**: 단항/이항/삼항 · 피연산자(operand) · 우선순위(산술>비교>논리)
- **복합문**: header `:` suite · 들여쓰기 · `if`/`elif`/`else` · `pass`
- **함수**: `def` · `return` · "호출식을 평가해야 값이 나온다"
- **알고리즘**: 최댓값(**챔피언 갱신 패턴**) · 중앙값(정의 → 효율) · *올바른 알고리즘 = 어떤 입력에도 결과가 같음*
- **순서도**: 타원(시작/끝)·직사각형(처리)·평행사변형(입출력)·마름모(판단)·미리 정의한 처리(함수 호출)


---
# Part 1. 변수 · 형변환


### 문제 1 🟢 [예측] — `input()`은 문자열을 준다

오늘 `int(input())`으로 감싼 이유를 확인해보자. 사용자가 `100`, `99`를 입력했다고 가정(값을 직접 넣음).
각 줄의 출력을 **예측**하고 실행해봐.

In [1]:
a = "100"   # input()이 준 값 (문자열!)
b = "99"

print(a > b)              # False
print(int(a) > int(b))    # True


# 한 줄 설명: 데이터 타입이 다르기 때문!

False
True


### 문제 2 🟢 [예측] — 문자열 vs 숫자 연산

같은 "30"처럼 보여도 따옴표 유무로 다르게 동작해.

In [2]:
print("30" + "5")   # 305
print(30 + 5)       # 35
print("30" * 3)     # 303030
print("30" + 5)   # ? 주석 풀면 무슨 일이? (왜?) : 오류가 나죠 문자열 + 숫자열 하면 큰일나죠

305
35
303030


TypeError: can only concatenate str (not "int") to str

### 문제 3 🟡 [설명] — 형변환과 진수

`int()`의 두 번째 인자는 "몇 진수로 읽을지"였어. 아래를 예측하고, 마지막은 **왜 에러인지** 설명해.

In [3]:
print(int('1010'))      # 1010
print(int('1010', 2))   # 1010을 2진수로.
print(int('FF', 16))    # FF를 16진수로
print(int('102', 2))  # ? 주석 풀면 에러. 왜일까? 오잉 진짜 왜지? 102를 2진수로... 2는 2진수에 없어요 ㅠ  0,1만 있어

1010
10
255


ValueError: invalid literal for int() with base 2: '102'

---
# Part 2. 출력 — `sep` / `end`


### 문제 4 🟢 [예측] — `sep`은 값 사이, `end`는 맨 끝

아래 블록의 **최종 출력 화면**을 정확히 예측해봐 (줄바꿈까지).

In [ ]:
print('A', end='')
print('B', end='-')
print('C')                  # \n돼서 나오겠지

print(2026, 7, 6, sep='-')  # 2026-7-6
print('x', 'y', 'z')        # 띄어쓰기

AB-C
2026-7-6
x y z


### 문제 5 🟡 [구현] — 의도대로 고치기

`이름:` 뒤에 **줄바꿈만 없애고** 입력을 바로 옆에서 받고 싶어. `end`에 뭘 넣어야 할까?

In [6]:
# 아래 ??? 를 올바른 값으로 고쳐보세요
name = "홍길동"  # input() 대신
print('이름:', end=' ')
print(name)
# 기대 출력(한 줄):  이름:홍길동

이름: 홍길동


---
# Part 3. 실수(float)의 함정


### 문제 6 🟢 [설명] — `0.1 + 0.2 == 0.3` 이 False?

실행하면 `False`가 나와. **왜인지 한 줄로** 설명해봐. (키워드: 2진수로 딱 안 떨어짐 → 근사값)

In [ ]:
print(0.1 + 0.2)
print(0.1 + 0.2 == 0.3) # 오 신기하다

0.30000000000000004
False


### 문제 7 🟡 [구현] — 실수는 이렇게 비교한다

실수는 `==`로 비교하면 안 돼. 오차 허용범위로 비교하는 함수를 완성해.

In [ ]:
def almost_equal(a, b):
    return abs(a - b) < 1e-9 # abs(a, b) a, b 두 값 사이의 절댓값
    #       힌트: abs(a - b)
    pass

print(almost_equal(0.1 + 0.2, 0.3))   # 기대: True
print(almost_equal(1.0, 2.0))         # 기대: False

True
False


### 문제 8 🔴 [예측] — 오차는 누적된다

`0.1`을 100번 더하면 정확히 `10.0`일까? 예측하고 실행해봐. 그리고 이게 왜 시뮬레이션·수치계산에서 중요한지 생각해봐.

In [ ]:
total = sum([0.1] * 100)
print(total)
print(total == 10.0)   # False

9.99999999999998
False


---
# Part 4. 연산자


### 문제 9 🟡 [설명] — 피연산자(operand) 세기

식 `a + b - c > d` 에 대해 답해봐:

1. `>` 연산자의 **왼쪽 피연산자**는 정확히 무엇? (`a`? 아니면?)
2. `-a`의 `-`와 `a - b`의 `-`는 각각 **단항/이항** 중 뭐?
3. 이 식 전체를 평가한 결과의 **타입**은? (숫자? bool?)

In [ ]:
a, b, c, d = 10, 5, 3, 4
result = a + b - c > d
print(result, type(result).__name__)   # 예측하고 확인

True bool


### 문제 10 🟢 [예측] — 우선순위: 산술 > 비교 > 논리

각 줄이 어떻게 묶여서 계산되는지 예측해봐.

In [ ]:
print(2 + 3 > 4)           # ? 어떻게 묶이나 (2, 3)
print(5 > 3 and 2 > 1)     # ? (5 > 3), (2>1)
print(1 + 1 == 2 and 3 > 5)  # ? 어떤 순서로?

True
True
False


### 문제 11 🟡 [구현] — 조건부 표현식 (삼항 연산자)

파이썬 유일의 삼항 연산자: `값1 if 조건 else 값2`. 이걸로 구현해봐.

In [ ]:
x, y = 7, 12

# x, y 중 큰 값을 bigger 에 (내장 max 쓰지 말고 조건부 표현식으로!)
bigger = None  # TODO
print(bigger)  # 기대: 12

---
# Part 5. 분기 · 복합문


### 문제 12 🟢 [구현] — 부호 판정 (3분기)

정수 `n`의 부호를 `'양수'`/`'음수'`/`'0'` 으로 돌려주는 함수를 완성해.

In [ ]:
def sign(n):
    # TODO: if / elif / else 로 3분기
    pass

print(sign(5))    # 양수
print(sign(-3))   # 음수
print(sign(0))    # 0

### 문제 13 🟡 [디버깅] — `else`가 없으면? `pass`는 뭐지?

오늘 `branch2.py`는 `else`가 없었어. `branch2(4)`를 실행하면 무슨 일이?

In [ ]:
def branch2(n):
    if n == 1:
        print('A')
    elif n == 2:
        print('B')
    elif n == 3:
        print('C')

branch2(1)
branch2(4)   # 화면에 뭐가 나올까? (또는 안 나올까?)

답해봐:
1. `branch2(4)`가 **아무것도 출력 안 하는 이유**는?
2. `else: pass` 의 `pass`는 무슨 뜻? **없으면** 어떻게 될까?
3. 🔴 아래처럼 `else:` 본문을 **비우면** 어떤 에러가, **언제**(실행 전/실행 중) 날까?

In [ ]:
# 🔴 예측만 해보기 (실행하면 에러)
# def broken(n):
#     if n == 1:
#         print('A')
#     else:
#         # (비어있음)
#     print('done')

### 문제 14 🔴 [설명] — 분기의 "완전성"

`sign` 함수의 3분기(`n>0` / `n<0` / `else`)는 **상호 배타적**이고 **완전**했어.

1. "상호 배타적"과 "완전(빠짐없음)"이 각각 무슨 뜻?
2. `else` 대신 `elif n == 0:` 으로 바꾸면 어떤 **위험**이 생길까? (`else`가 보장하는 것)

---
# Part 6. 함수 (def)


### 문제 15 🟢 [예측] — `return` vs `print`

"함수 호출식을 평가해야 반환값을 얻는다"가 오늘 핵심 문장. 두 함수의 차이를 예측해봐.

In [ ]:
def add_return(a, b):
    return a + b

def add_print(a, b):
    print(a + b)      # return이 아니라 print

r1 = add_return(3, 4)
r2 = add_print(3, 4)

print("r1 =", r1)     # ?
print("r2 =", r2)     # ? 왜 이 값일까

### 문제 16 🔴 [디버깅] — 내장 이름 가리기(shadowing)

오늘 함수 이름을 `max`로 지었지. 이게 왜 위험한지 재현해봐.

In [ ]:
print(max([3, 1, 2]))   # 내장 max: 리스트 최댓값 → 잘 됨

def max(a, b, c):       # ← 내장 max를 덮어씀
    m = a
    if b > m: m = b
    if c > m: m = c
    return m

print(max(3, 1, 2))     # 내가 만든 버전: 잘 됨
# print(max([3, 1, 2])) # 주석 풀면? 무슨 에러가 왜?

# 질문: 함수 이름을 뭐로 바꾸면 안전할까? (2가지 이상)

---
# Part 7. 알고리즘 — 최댓값 · 중앙값
> **오늘의 메인.** 여기가 자료구조 스터디의 진짜 시작점.


### 문제 17 🟢 [구현] — 챔피언 갱신 패턴

`max3`의 핵심 패턴: **"첫 값을 임시 챔피언으로 두고, 하나씩 비교하며 갱신"**.
이 패턴 그대로 **최솟값** 함수를 완성해.

In [ ]:
def min3(a, b, c):
    m = a           # a를 임시 챔피언으로
    # TODO: b, c와 비교하며 갱신 (부등호 방향 주의!)
    return m

print(min3(3, 1, 2))   # 기대: 1
print(min3(5, 5, 9))   # 기대: 5

### 문제 18 🟡 [구현] — 중앙값을 3가지 방법으로

같은 문제를 **추상화 레벨을 바꿔가며** 풀어보는 게 오늘의 핵심 훈련.

In [ ]:
# (1) 🟡 정렬에 위임 — 한 줄로
def median_sort(a, b, c):
    # TODO: sorted([...]) 의 가운데(인덱스 1)
    pass

# (2) 🟡 경계값 논리 — 연쇄 비교
#     "자기 자신이 나머지 둘 사이에 있으면 중앙값"
def median_chain(a, b, c):
    # TODO: c <= a <= b  같은 연쇄 비교 사용
    #       (a가 사이) or (b가 사이) or else c
    pass

# (3) 🔴 min/max 트릭 — 합에서 최대·최소를 빼면 가운데만 남는다
def median_trick(a, b, c):
    # TODO
    pass

tests = [(1,2,3), (3,1,2), (2,2,1), (5,5,5)]
for f in (median_sort, median_chain, median_trick):
    print(f.__name__, [f(*t) for t in tests])
    # 기대: 모두 [2, 2, 2, 5]

### 문제 19 🔴 [분석] — "정답"과 "효율적인 답"은 다르다

오늘 네가 만든 `median3a`(`and` 버전)는 **정답이지만** 책이 "비효율적"이라 했어.

1. **어디서 중복 비교**가 일어나는지 짚어봐. (첫 `if`와 `elif`에서 같은 쌍을 또 비교)
2. 중첩 if 버전(`median3.py`)은 왜 비교를 덜 하나? ("이미 한 비교 결과를 재사용")
3. 값이 3개일 땐 중첩 if가, 값이 많아지면 정렬이 유리해지는 이유는?

> 💡 이 "중복 비교 제거 = 이미 아는 정보 재사용"이 알고리즘 최적화의 가장 기본 감각. 앞으로 계속 나와.

---
# Part 8. 순서도 ↔ 코드


### 문제 20 🟢 [설명] — 도형 매칭

아래 각 코드 줄이 순서도의 **어떤 도형**인지 짝지어봐.

후보: **타원(시작/끝)** · **직사각형(처리)** · **평행사변형(입출력)** · **마름모(판단)** · **미리 정의한 처리(함수 호출)**

```python
a = int(input())         # (가)
maximum = a              # (나)
if b > maximum:          # (다)
result = median(a,b,c)   # (라)
print(result)            # (마)
```


### 문제 21 🟡 [구현] — 순서도 → 코드

아래 순서도를 파이썬 코드로 옮겨봐.

```
[시작] → [정수 n 입력] → ◇ n을 2로 나눈 나머지 == 0 ? ◇
                              │예              │아니오
                        ["짝수" 반환]     ["홀수" 반환]
                                    → [끝]
```
힌트: 나머지 연산자 `%` 사용 (`n % 2`)

In [ ]:
def even_odd(n):
    # TODO
    pass

print(even_odd(4))   # 기대: 짝수
print(even_odd(7))   # 기대: 홀수

### 문제 22 🔴 [설명] — 도형의 경계

1. **"처리(직사각형)"** 와 **"미리 정의한 처리"** 의 차이를 한 문장으로.
2. **"데이터(평행사변형=입출력)"** 와 **"처리(직사각형)"** 를 가르는 기준 한 문장.
   (힌트: 값이 프로그램 경계를 넘나드는가?)

---
# ✅ 정답 & 해설

> 먼저 스스로 풀었지? 답만 보지 말고 **"왜"** 를 꼭 읽자.


### Part 1. 변수 · 형변환
**1** 🟢 — `a > b` → **`False`** (문자열은 **사전순** 비교: `'1'` < `'9'`). `int(a) > int(b)` → **`True`**.
→ `input()`은 항상 `str`이라 숫자처럼 보여도 문자 비교됨. 그래서 `int()`로 감싼다.

**2** 🟢 — `"305"`, `35`, `"303030"`. `"30" + 5`는 **TypeError**(str+int 불가).
→ 문자열 `+`=이어붙이기 `*`=반복, 숫자는 산술.

**3** 🟡 — `10`, `10`, `255`. `int('102', 2)`는 **에러**: 2진수엔 숫자 `2`가 없음(0,1만). 진수 범위를 벗어난 문자.

### Part 2. 출력
**4** 🟢 —
```
AB-C          (첫 줄: 'A'+'B'+'-'+'C'+개행)
2026-7-6
x y z         (sep 기본값 = 공백)
```
**5** 🟡 — `end=''` (빈 문자열). 다른 문자를 넣으면 그게 뒤에 붙어버림.

### Part 3. float
**6** 🟢 — `0.30000000000000004`, `False`.
→ `0.1`은 2진수로 딱 안 떨어져서(무한 반복) 저장 순간 근사값이 됨. 그 오차 때문에 `0.3`과 미세하게 어긋남.

**7** 🟡 —
```python
def almost_equal(a, b):
    return abs(a - b) < 1e-9
```

**8** 🔴 — `9.99999999999998`, `== 10.0` 은 **False**. 오차가 100번 누적됨.
→ 시뮬레이션·수치계산에서 float 비교를 `==`로 하면 안 되는 이유. (너 MuJoCo/JAX에서 만날 문제.)

### Part 4. 연산자
**9** 🟡 —
1. `>`의 왼쪽 피연산자는 `a`가 아니라 **`(a + b - c)`라는 계산 결과** 하나.
2. `-a`의 `-`=**단항**(부호 반전), `a - b`의 `-`=**이항**(뺄셈).
3. 결과 타입 = **`bool`**.

**10** 🟢 — `True`(=`(2+3)>4`=`5>4`), `True`, `False`(=`True and False`).
→ 산술이 먼저 묶이고, `and`는 맨 나중.

**11** 🟡 — `bigger = x if x > y else y` → `12`.

### Part 5. 분기 · 복합문
**12** 🟢 —
```python
def sign(n):
    if n > 0:   return '양수'
    elif n < 0: return '음수'
    else:       return '0'
```

**13** 🟡🔴 —
1. `branch2(4)`: 어떤 조건에도 안 걸리고 `else`도 없어서 **아무것도 출력 안 함**(그냥 지나감).
2. `pass` = "아무것도 안 하고 지나가라". 문법상 **본문이 비면 안 되는데** 할 일이 없을 때 자리 채우기. 없으면 `IndentationError`.
3. 🔴 `else:` 본문을 비우면 → **`IndentationError`**, **실행 전**(문법 검사 단계)에 발생.

**14** 🔴 —
1. **상호 배타적** = 한 번에 한 분기만 참(겹치지 않음). **완전** = 모든 경우가 어느 분기엔 반드시 속함.
2. `elif n == 0`으로 바꾸면 정수에선 결과 같지만, **완전성이 깨질 위험**. `else`는 "나머지 전부"를 보장해서 빠지는 경우가 없음.

### Part 6. 함수
**15** 🟢 — `r1 = 7`. `add_print`는 `return`이 없어서 **`None` 반환** → `r2 = None`. (7은 화면엔 찍히지만 값으로 안 돌아옴.)

**16** 🔴 — 마지막 줄은 **TypeError**: 내 `max`는 인자 3개를 요구하는데 리스트 1개를 줘서. 내장 `max`가 가려짐(shadowing).
→ 안전한 이름: `max3`, `my_max`, `get_max` 등. (되돌리려면 `del max`.)

### Part 7. 알고리즘
**17** 🟢 —
```python
def min3(a, b, c):
    m = a
    if b < m: m = b
    if c < m: m = c
    return m
```

**18** 🟡🔴 —
```python
def median_sort(a, b, c):
    return sorted([a, b, c])[1]

def median_chain(a, b, c):
    if c <= a <= b or b <= a <= c: return a
    elif c <= b <= a or a <= b <= c: return b
    else: return c

def median_trick(a, b, c):
    return a + b + c - max(a, b, c) - min(a, b, c)
```

**19** 🔴 —
1. `median3a`는 첫 `if`에서 `a`와 `b`를 비교하고, `elif`에서 `b`와 `a`를 **또** 비교 → 같은 쌍 중복 검사.
2. 중첩 if는 `a >= b`가 참임을 알면 그 사실을 **재사용**해 `b`와 `c`만 더 봄 → 비교 최소.
3. 3개는 몇 번 비교로 끝(정렬은 오버헤드). 개수가 늘면 손으로 짠 분기는 확장 불가, 정렬은 `sorted(...)[len//2]`로 일반화.

### Part 8. 순서도
**20** 🟢 — (가) 평행사변형(입력) · (나) 직사각형(처리) · (다) 마름모(판단) · (라) 미리 정의한 처리 · (마) 평행사변형(출력)

**21** 🟡 —
```python
def even_odd(n):
    if n % 2 == 0:
        return "짝수"
    else:
        return "홀수"
```

**22** 🔴 —
1. 처리 = 그 자리에서 직접 계산/대입. 미리 정의한 처리 = **딴 데 정의된 함수를 호출**.
2. 값이 프로그램 **경계를 넘나들면(입출력)** 데이터(평행사변형), **안에서 굴리면** 처리(직사각형).


---
## 🎯 오늘의 핵심 3줄
1. **`input()`은 str** — 숫자로 쓰려면 `int()`. 실수는 근사값이라 `==` 금지.
2. **분기의 정확성 = 완전성 + 상호배타**, 그다음 단계가 **중복 비교 제거(효율)**.
3. **함수 = "로직에 이름 붙여 빼두기"**, 호출식을 평가해야 값이 나온다. (순서도의 "미리 정의한 처리")

### 📅 스터디 진행 가이드
- 🟢 (전원): 1, 2, 4, 6, 10, 12, 15, 17, 20
- 🟡 (목표): 3, 5, 7, 9, 11, 13, 18, 21
- 🔴 (도전): 8, 14, 16, 19, 22

> 다음 2일차 노트북 맨 앞엔 **[Remind]** 섹션으로 오늘 핵심을 한 단계 높여 다시 풀 예정. 🔥
